pkg/runtime/dynamicload.go

```diff
@@ -0,0 +1,97 @@
+package runtime
+
+import (
+	"path/filepath"
+
+	"github.com/dapr/dapr/pkg/components"
+	"github.com/fsnotify/fsnotify"
+
+	componentsV1alpha1 "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
+)
+
+func (a *DaprRuntime) watchPathForDynamicLoading() {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		log.Errorf("unable to create watcher for dynamic components, dynamic loading will not be supported: %s", err)
+	}
+	defer watcher.Close()
+	err = watcher.Add(a.runtimeConfig.Standalone.ComponentsPath)
+	if err != nil {
+		log.Error(err)
+	}
+	for {
+		select {
+		case event, ok := <-watcher.Events:
+			if !ok {
+				return
+			}
+			log.Debug("observed a file event in components directory:", event)
+			if event.Op == fsnotify.Create || event.Op == fsnotify.Write {
+				err := a.loadDynamicComponents(event.Name)
+				if err != nil {
+					log.Errorf("failed to load components from file: %s err: %s", event.Name, err)
+				}
+			} else if event.Op == fsnotify.Remove || event.Op == fsnotify.Rename {
+				for _, comp := range a.dynamicComponents[DynamicComponentsManifest(event.Name)] {
+					log.Warnf("file: %s deleted, component: %s will not be loaded on sidecar restart.", filepath.Base(event.Name), comp.Name)
+				}
+			}
+		case err, ok := <-watcher.Errors:
+			if !ok {
+				return
+			}
+			log.Errorf("error while watching components directory for file events: %s", err)
+		}
+	}
+}
+
+func (a *DaprRuntime) loadDynamicComponents(manifestPath string) error {
+	loader := components.NewStandaloneComponents(a.runtimeConfig.Standalone)
+	log.Info("found new file in components directory, loading components from file: %s", manifestPath)
+	fileComps := loader.LoadComponentsFromFile(manifestPath)
+	componentsToLoad := []componentsV1alpha1.Component{}
+
+	for _, comp := range fileComps {
+		if a.IsComponentLoaded(comp) {
+			log.Infof("component is already loaded, skipping. name: %s, type: %s/%s", comp.ObjectMeta.Name, comp.Spec.Type, comp.Spec.Version)
+			continue
+		}
+		log.Infof("found component. name: %s, type: %s/%s", comp.ObjectMeta.Name, comp.Spec.Type, comp.Spec.Version)
+		componentsToLoad = append(componentsToLoad, comp)
+	}
+
+	a.dynamicComponents[DynamicComponentsManifest(manifestPath)] = componentsToLoad
+
+	authorizedComps := a.getAuthorizedComponents(componentsToLoad)
+	a.componentsLock.Lock()
+	a.components = append(a.components, authorizedComps...)
+	a.componentsLock.Unlock()
+
+	for _, comp := range authorizedComps {
+		err := a.processComponentAndDependents(comp)
+		if err != nil {
+			if !comp.Spec.IgnoreErrors {
+				log.Warnf("error processing component %s, error: %s", comp.Name, err.Error())
```

Comment:
The intention was that daprd should not stop if user is trying to load component dynamically and there is some error in the manifest, giving a chance to rectify the issue without a sidecar restart.

Reply:
The intent behind `ignoreErrors` field is lost in the case of dynamic components then. We are saying that even if there is some error in loading a dynamic component, then `daprd` will continue on as such. 

@dapr/maintainers-dapr Is this the intended flow for dynamic components?

Reply:
Then again, not restarting the sidecar would make the error in the new dynamically loaded component go unnoticed in most cases, which is what `ignoreErrors` is trying to solve for mainly. I recommend to honor `ignoreErrors` for dynamically loaded components.

Reply:
@yaron2 / @mukundansundar - for a daprd startup ignore errors indicates that all the configuration SHOULD be loaded else, dapr should not continue to run. Hence the implementation to flag loading failures as critical and trigger a restart, which is the most we can do to initiate recovery.
For dynamic loading however, the intent is to minimise disruption to the running setup. So user is able to correct the error and retry the loading. Since the loading is halted at the initial phase, there is no impact on the rest of the daprd. Restart would not add any value since we would technically be bringing Daprd back to the state that it was in before restart. Also, some configuration needs to be redone, for example subscription to configuration via the config building block. Hence the approach to log a proper error and allow daprd to function un-interrupted. 
So ignore errors would be true always for dynamic loading.

Reply:
@yaron2 -  Exploring options to make the loading error visible for action - One thought is to be able to surface errors to the app instead of restarting the sidecar, we could send a notification on the app channel with a relevant error code and the detailed failure reason. This would bring attention on the failure. So, if ignoreErrors is set to false, a notification can be sent to the app channel when a dynamically loaded component fails during load. Any thoughts on this?

Reply:
@akhilac1 was the intent for this feature to be used primarily for development (so was my understanding) or for production? In the first case, asking the developer to add a route to the app, seems especially an overkill, as they'd also be responsible for removing the route in production.

Reply:
There are two options that I can see here. 

One we continue to honor `ignoreErrors` and that will disrupt sidecar runs for any dynamic component that fails to initialize. This way `ignoreErrors` as a property will be consistent in definition and use across both normal components and dynamically loaded components. 

The other option is that we ignore the `ignoreErrors` field to reduce disruptions to sidecar. But in this case we need a way to surface the component init errors to the app/user. **Error logs are not sufficient for that.** Even if we have a way to notify the app that the dynamic load of a component failed, we also need to clearly document the behavioral differences between `ignoreErrors` field when used with a normal component vs when used with a pluggable component. 

IMO I believe that both options offers different benefits. The first option seems to have a lesser learning curve and lesser implementation change wrt to the app. Also it follows the expected behavior when the field is configured. 
While in the second option, it offers uninterrupted run except for the particular features that the dynamically loaded component alone targets. But it also has different behaviors for a field in different modes, and might potentially lead to code change in the app to consume any component init errors. 

I think we should continue to honor `ignoreErrors` field so that there are no two different behaviors in two different modes. We should maintain the consistency.

Reply:
I agree. Hot reloading is meant to allow users to load components without restarting the sidecar. It says nothing about changing the behavior when encountering errors. If a hot reload event occured, at that point this feature ends and the regular lifecycle of component validation begins.